flight/Modules/VtolPathFollower/vtol_follower_control.c

```diff
@@ -72,123 +197,184 @@ int32_t vtol_follower_control_path(const float dT, const PathDesiredData *pathDe
-	velocityDesired.North += progress->correction_direction[0] * error_speed * scale;
-	velocityDesired.East += progress->correction_direction[1] * error_speed * scale;
+
+	/* Sum the desired path movement vector with the correction vector */
+	commands_ned[0] = progress->path_direction[0] * groundspeed +
+	    progress->correction_direction[0] * error_speed;
 	
-	// Interpolate desired velocity along the path
-	float altitudeSetpoint = pathDesired->Start[2] + (pathDesired->End[2] - pathDesired->Start[2]) *
-	    bound_min_max(progress->fractional_progress, 0, 1);
+	commands_ned[1] = progress->path_direction[1] * groundspeed +
+	    progress->correction_direction[1] * error_speed;
 
-	float downError = altitudeSetpoint - positionActual.Down;
-	velocityDesired.Down = pid_apply_antiwindup(&vtol_pids[DOWN_POSITION], downError,
-		-guidanceSettings.VerticalVelMax, guidanceSettings.VerticalVelMax, dT);
+	/* Limit the total velocity based on the configured value. */
+	vtol_limit_velocity(commands_ned, guidanceSettings.HorizontalVelMax);
+
+	VelocityDesiredData velocityDesired;
+
+	velocityDesired.North = commands_ned[0];
+	velocityDesired.East = commands_ned[1];
+	velocityDesired.Down = commands_ned[2];
 
 	VelocityDesiredSet(&velocityDesired);
 
 	return 0;
 }
 
-/**
- * Control algorithm to stay or approach at a fixed location.
- * @param[in] dT the time since last evaluation
- * @param[in] ned The position to attempt to hold
- * This method does not attempt any particular path, simply a straight line
- * approach. The calculated velocity to attempt is stored in @ref 
- * VelocityDesired. 
- *
- * Takes in @ref PositionActual and compares it to @ref PositionDesired 
- * and computes @ref VelocityDesired
- */
-int32_t vtol_follower_control_endpoint(const float dT, const float *hold_pos_ned)
-{
+static int32_t vtol_follower_control_simple(const float dT,
+	const float *hold_pos_ned, bool landing, bool update_status) {
 	PositionActualData positionActual;
 	VelocityDesiredData velocityDesired;
 	
 	PositionActualGet(&positionActual);
-	VelocityDesiredGet(&velocityDesired);
-	
-	float northError;
-	float eastError;
-	float downError;
-	float northCommand;
-	float eastCommand;
 	
-	const float cur_pos_ned[3] = { positionActual.North, positionActual.East, positionActual.Down };
+	VelocityActualData velocityActual;
+
+	VelocityActualGet(&velocityActual);
+
+	/* Where would we be in ___ second at current rates? */	
+	const float cur_pos_ned[3] = {
+		positionActual.North +
+		    velocityActual.North*guidanceSettings.PositionFeedforward,
+		positionActual.East +
+		    velocityActual.East*guidanceSettings.PositionFeedforward,
+		positionActual.Down };
+
+	float errors_ned[3];
+
+	vtol_calculate_distances(cur_pos_ned, hold_pos_ned, errors_ned, 0);
+
+	float horiz_error_mag = vtol_magnitude(errors_ned, 2);
+	float scale_horiz_error_mag;
+
+	if (horiz_error_mag > 0.00001f) {
+		scale_horiz_error_mag = vtol_deadband(horiz_error_mag,
```

Comment:
can you explain what this math is doing (comments?). I think you can make this much cleaner if you make ```vtol_calculate_distances``` return the norm when normalizing (which it computed anyway). then you can do this

float horiz_error_mag = vtol_calculate_distances(cur_pos_ned, hold_pos_ned, errors_ned, false);
horiz_error_mag = vtol_deadband(horiz_error_mag, guidanceSettings.EndpointDeadbandWidth guidanceSettings.EndpointDeadbandCenterGain, vtol_end_m, vtol_end_r);
float damped_ne[2] = { errors_ned[0] * horiz_error_mag, errors_ned[1] * horiz_error_mag };